File: backend/app/models/progress.py
"""Pydantic models for the progress tracking API.

See docs/superpowers/specs/2026-05-03-phase-1-progress-tracking-design.md
for the design rationale.
"""

from datetime import datetime
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# Canonical section keys — every unit currently has exactly these five.
# Mirrored at the DB layer via a CHECK constraint (migration
# 20260504000002_constrain_progress_keys.sql) and at the frontend in
# `src/features/lessons/lesson.types.ts:SectionKey`. Adding a new section
# type is a deliberate three-place change.
SectionKey = Literal["overview", "grammar", "vocabulary", "dialogues", "activities"]

# Unit slug pattern: matches "unit-1", "unit-12", etc. Tightening here
# stops typos and arbitrary writes from polluting the projection tables.
_UNIT_SLUG_PATTERN = r"^unit-\d+$"

# Exercise IDs are author-defined strings like "mc-1" or "fb-3a". Lock
# down to lowercase kebab-case + digits with a sane max length.
_EXERCISE_ID_PATTERN = r"^[a-z0-9][a-z0-9-]*$"


# ----- Request models -----

class LessonSectionCompletedRequest(BaseModel):
    unit_slug: str = Field(pattern=_UNIT_SLUG_PATTERN)
    section_key: SectionKey


class ExerciseAttemptRequest(BaseModel):
    unit_slug: str = Field(pattern=_UNIT_SLUG_PATTERN)
    section_key: SectionKey
    exercise_id: str = Field(min_length=1, max_length=64, pattern=_EXERCISE_ID_PATTERN)
    is_correct: bool


class FlashcardReviewRequest(BaseModel):
    flashcard_id: UUID
    status: Literal["known", "unknown"]


# ----- Single-write response models -----

class LessonSectionProgressResponse(BaseModel):
    unit_slug: str
    section_key: str
    completed_at: datetime


class ExerciseAttemptResponse(BaseModel):
    id: int
    attempted_at: datetime


class FlashcardReviewResponse(BaseModel):
    id: int
    reviewed_at: datetime


# ----- Summary response (composite read) -----

class CompletedSection(BaseModel):
    unit_slug: str
    section_key: str
    completed_at: datetime


class AttemptsSummary(BaseModel):
    total: int
    correct: int


class FlashcardsSummary(BaseModel):
    reviewed_total: int
    currently_known: int


class StreakSummary(BaseModel):
    current_days: int


class ActivityCounts(BaseModel):
    """Pre-computed counts the dashboard renders directly. Some fields
    duplicate values from the breakdown blocks above — intentional, so
    the dashboard can render `summary.activity` without recomputing.
    """
    lessons_completed: int        # count of FULLY-completed units
    exercises_attempted: int      # mirrors AttemptsSummary.total
    exercises_correct: int        # mirrors AttemptsSummary.correct
    flashcards_reviewed: int      # mirrors FlashcardsSummary.reviewed_total
    flashcards_mastered: int      # mirrors FlashcardsSummary.currently_known


class ProgressSummaryResponse(BaseModel):
    sections_completed: List[CompletedSection]
    exercise_attempts: AttemptsSummary
    flashcards: FlashcardsSummary
    streak: StreakSummary
    study_days_this_week: int
    last_active_at: Optional[datetime]
    activity: ActivityCounts

File: src/lib/api/progress.ts
import { authedGet, authedPostJson } from "@/lib/api/authedFetch";

export type CompletedSection = {
  unit_slug: string;
  section_key: string;
  completed_at: string;
};

export type ProgressSummary = {
  sections_completed: CompletedSection[];
  exercise_attempts: { total: number; correct: number };
  flashcards: { reviewed_total: number; currently_known: number };
  streak: { current_days: number };
  study_days_this_week: number;
  last_active_at: string | null;
  activity: {
    lessons_completed: number;
    exercises_attempted: number;
    exercises_correct: number;
    flashcards_reviewed: number;
    flashcards_mastered: number;
  };
};

class ProgressAPIClass {
  async completeSection(args: { unitSlug: string; sectionKey: string }) {
    try {
      return await authedPostJson<CompletedSection>(
        "/me/progress/complete-section",
        { unit_slug: args.unitSlug, section_key: args.sectionKey },
      );
    } catch (err) {
      console.error("ProgressAPI.completeSection failed", err);
      return null;
    }
  }

  async attemptExercise(args: {
    unitSlug: string;
    sectionKey: string;
    exerciseId: string;
    isCorrect: boolean;
  }) {
    try {
      return await authedPostJson<{ id: number; attempted_at: string }>(
        "/me/progress/attempt-exercise",
        {
          unit_slug: args.unitSlug,
          section_key: args.sectionKey,
          exercise_id: args.exerciseId,
          is_correct: args.isCorrect,
        },
      );
    } catch (err) {
      console.error("ProgressAPI.attemptExercise failed", err);
      return null;
    }
  }

  async reviewFlashcard(args: { flashcardId: string; status: "known" | "unknown" }) {
    try {
      return await authedPostJson<{ id: number; reviewed_at: string }>(
        "/me/progress/review-flashcard",
        { flashcard_id: args.flashcardId, status: args.status },
      );
    } catch (err) {
      console.error("ProgressAPI.reviewFlashcard failed", err);
      return null;
    }
  }

  getSummary() {
    // Read method: lets errors propagate so the hook can render an error state.
    return authedGet<ProgressSummary>("/me/progress/summary");
  }
}

export const ProgressAPI = new ProgressAPIClass();
